### PR TITLE
Let rec patterns

### PR DESCRIPTION
**Pull request by jneem**
Depends on #2010 

Allows `let rec` bindings to contain patterns. Also fixes a bug in lsp completions, where pattern bindings were accidentally already treated as recursive.

The destructuring is no longer desugared to a match block (which allows for slightly better diagnostics), but it is still as eager as a match block destructuring and it still uses the same pattern compilation under the hood. The motivations for destructuring eagerly,

- to share code with pattern compilation for match bindings,
- to have similar behavior in match blocks and let bindings, and
- to catch errors earlier (with lazy destructuring, `let 'Foo = 'Bar in true` won't error)

### CODE REVIEW

**Comment by yannham** [comment]
@jneem once it's rebased could please tag me, and I'll take a look!

**Comment by jneem** [comment]
@yannham it's ready for a look

File: core/src/typecheck/pattern.rs [hunk 1]
```diff
@@ -176,9 +176,9 @@ pub(super) trait PatternTypes {
     /// value will be checked against the type generated by `pattern_type` and the bindings will be
     /// added to the type environment.
     ///
-    /// The type of each "leaf" identifier will be assigned based on the `mode` argument. The
-    /// current possibilities are for each leaf to have type `Dyn`, to use an explicit type
-    /// annotation, or to be assigned a fresh unification variable.
+    /// The type of each "leaf" identifier will be assigned based on the `mode`
+    /// argument. The current possibilities are for each leaf to have type
+    /// `Dyn`, or to be assigned a fresh unification variable.
```
[the remark this answers is not in the document]
Comment (yannham) [reply core/src/typecheck/pattern.rs:181]: I think the third case is something like `match x { {foo : Number, bar: String} => ...}`, where we take the type from the annotation instead. Or at least so I recall.

File: core/src/transform/desugar_destructuring.rs [hunk 2]
```diff
@@ -48,63 +55,116 @@ pub fn desugar_fun(mut pat: Pattern, body: RichTerm) -> Term {
...
-        bound.push(rt);
+    // Outer bindings are the ones we called %b1 and %b2, and %empty_record_id in the doc above.
+    let mut outer_bindings = SmallVec::new();
+    // Mid bindings are the ones we called %r1 and %r2 above.
+    let mut mid_bindings = SmallVec::new();
+    // Inner bindings are the ones that bind the actual variables defined in the patterns.
+    let mut inner_bindings = SmallVec::new();
+    let mut error_tests = Vec::new();
+
+    let empty_record_id = LocIdent::fresh();
```
[the remark this answers is not in the document]
Comment (yannham) [reply core/src/transform/desugar_destructuring.rs:108]: Ah, right, the interface is to take an ident :+1:

File: core/src/transform/desugar_destructuring.rs [hunk 3]
```diff
@@ -48,63 +55,116 @@ pub fn desugar_fun(mut pat: Pattern, body: RichTerm) -> Term {
...
+        .into_iter()
+        .rev()
+        .fold(body, |acc, (check, error)| {
+            make::if_then_else(check, error, acc)
+        });
+
+    let attrs = LetAttrs {
+        binding_type: BindingType::Normal,
+        rec,
+    };
```
[the remark this answers is not in the document]
Comment (jneem) [reply core/src/transform/desugar_destructuring.rs:147]: This one has a `rec`...

**Pull request merged**